Leanplum-SDK/Classes/LPDatabase.m

```diff
@@ -0,0 +1,226 @@
+        willSendErrorLog = YES;
+        NSException *exception = [NSException exceptionWithName:errorName
+                                                         reason:reason
+                                                       userInfo:nil];
+        leanplumInternalError(exception);
+    }
+    
+    // If SQLite is corrupted create a new one.
+    // Using retryOnCorrupt to prevent infinite loop.
+    if (result == SQLITE_CORRUPT || !retryOnCorrupt) {
+        [[NSFileManager defaultManager] removeItemAtPath:[LPDatabase sqliteFilePath] error:nil];
+        retryOnCorrupt = YES;
+        [self initSQLite];
+    }
+}
+
+/**
+ * Helper method that returns sqlite statement from query.
+ * Used by both runQuery: and rowsFromQuery.
+ */
+- (sqlite3_stmt *)sqliteStatementFromQuery:(NSString *)query
+                               bindObjects:(NSArray *)objectsToBind
+{
+    if (!query) {
+        return nil;
+    }
+    
+    sqlite3_stmt *statement;
+    int __block result = sqlite3_prepare_v2(sqlite, [query UTF8String], -1, &statement, NULL);
+    if (result != SQLITE_OK) {
+        LPLog(LPError, @"Preparing '%@': %s (%d)", query, sqlite3_errmsg(sqlite),
+              result);
+        return nil;
+    }
+    
+    // Bind objects.
+    // It is recommended to use this instead of making a full query in NSString to
+    // prevent from SQL injection attacks and errors from having a quotation mark in text.
+    [objectsToBind enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
+        if (![obj isKindOfClass:[NSString class]]) {
+            LPLog(LPError, @"Bind object have to be NSString.");
+        }
+        
+        result = sqlite3_bind_text(statement, (int)idx+1, [obj UTF8String],
+                                   (int)[obj lengthOfBytesUsingEncoding:NSUTF8StringEncoding],
+                                   SQLITE_TRANSIENT);
+        
+        if (result != SQLITE_OK) {
+            LPLog(LPError, @"Binding %@ to %ld: %s (%d)", obj, idx+1, sqlite3_errmsg(sqlite),
+                  result);
+        }
+    }];
+    
+    return statement;
+}
+
+- (void)runQuery:(NSString *)query
+{
+    [self runQuery:query bindObjects:nil];
+}
+
+- (void)runQuery:(NSString *)query bindObjects:(NSArray *)objectsToBind
+{
+    @synchronized (self) {
+        sqlite3_stmt *statement = [self sqliteStatementFromQuery:query bindObjects:objectsToBind];
+        if (!statement) {
+            return;
+        }
+        
+        int result = sqlite3_step(statement);
+        if (result != SQLITE_DONE) {
+            [self handleSQLiteError:@"SQLite fail to run query" errorResult:result query:query];
+        }
+        willSendErrorLog = NO;
+        sqlite3_finalize(statement);
+    }
+}
+
+- (NSArray *)rowsFromQuery:(NSString *)query
+{
+    return [self rowsFromQuery:query bindObjects:nil];
+}
+
+- (NSArray *)rowsFromQuery:(NSString *)query bindObjects:(NSArray *)objectsToBind
+{
+    @synchronized (self) {
```

Comment:
You are not suppose to have sqlite access from different threads (it causes errors). I think it's safer to wrap the whole thing. Let me know if you have other suggestions.